# Used car dataset from eBay Kleinanzeigen, a classifieds section of the German eBay website.


The columns of dataset specify(description) following fields -

**dateCrawled** - When this ad was first crawled. All field-values are taken from this date.

**name** - Name of the car.

**seller** - Whether the seller is private or a dealer.

**offerType** - The type of listing.

**price** - The price on the ad to sell the car.

**abtest** - Whether the listing is included in an A/B test.

**vehicleType** - The vehicle Type.

**yearOfRegistration** - The year in which the car was first registered.

**gearbox** - The transmission type.

**powerPS** - The power of the car in PS.

**model** - The car model name.

**kilometer** - How many kilometers the car has driven.

**monthOfRegistration** - The month in which the car was first registered.

**fuelType** - What type of fuel the car uses.

**brand** - The brand of the car.

**notRepairedDamage** - If the car has a damage which is not yet repaired.

**dateCreated** - The date on which the eBay listing was created.

**nrOfPictures** - The number of pictures in the ad.

**postalCode** - The postal code for the location of the vehicle.

**lastSeenOnline** - When the crawler saw this ad last online.




We will now analyse this dataset to find useful insights about the given dataset


In [1]:
#importing useful libraries to analyse data
import pandas as pd
import numpy as np

#Reading autos.csv file 
autos = pd.read_csv("autos.csv",encoding="Latin-1")
# print(autos)

#print info. about converted dataframe
print(autos.info())
print(autos.head(3))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

From above experiment we can observe that our dataset contains 50000 rows & 20 distinct cols most of them which are of string type.
Datasets contains few null values as well which we will analyse more thoroughly later in the project.
One more thing to notice here is the cols name are not in python's snake_case,instead they are in camel case
Therefore we'll now convert some cols name to some relevant case & meaning

In [2]:
# converting cols name 
cols = autos.columns
cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.columns = cols
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_PS   model   odometer  registration_month fuel_type    brand  \
0       158  andere  150,000km                   3       lpg  peugeot   
1       286     7er  150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-26 00:00:00           0        79588   
1              nein  2016-04-04 00:00:00           0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

As we can see, we change the column names to make it more self explanatory and in python's standarized snake_case manner


# Further exploration of dataset

In [3]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

From above dataframe we can observe that two cols contains mostly the same values and those are:-
1) Seller
2) offer_type

The column that seems to need more investigation is:-
num_photos
So let's analyse more

In [4]:
#Analysing num_photos column
print(autos["num_photos"].unique())
print(autos["num_photos"].head(10))




[0]
0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: num_photos, dtype: int64


We can conclude that num_photos col contains only 1 value i.e '0' 
We will now remove the above mentioned three cols from our dataset


In [5]:
autos.drop(['seller', 'offer_type', 'num_photos'], axis=1 , inplace = True)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

Columns that needs to be clean & converted for further analysis
1) price
2) odometer

In [6]:
#Removing character from column values & convert their datatypes
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","").astype(int)
autos["odometer"] =autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

#Renaming column
autos.rename({"odometer" :"odometer_km"}, axis='columns',inplace=True)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

Further Exploring the data.
Remove outliers if any


In [7]:
print("Price information")
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().sort_index().head(25))
print(autos["price"].value_counts().sort_index(ascending = False).head(20))

print("Kilometer information")
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts())


Price information
(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
Name: price, dtype: int64
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64
Kilometer information
(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.0000

With above stats, we can conclude few things:-
**Kms**
1) Max km - 150000kms - freq(32424)
2) Min km - 5000kms -freq(967)

So this assumption looks fine as no value looks like an outlier

**price**
3) Max Price - $99999999 
4) Min Price - $0
Both observations sounds unreasonable
so,we'll restrict our observation in between $1 to $350000
values above $1 = 1421
values below $350000 = 14
Let's remove some rows acc to above conclusions
 
 

In [8]:
#Removing rows(outliers)

autos = autos[autos["price"].between(1,350001)]
print(autos["price"].shape)
autos["price"].describe()

(48565,)


count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Working with dates

There are a number of columns with date information:

date_crawled
registration_month
registration_year
ad_created
last_seen

The registration_month & registration_year are represented as numeric values therefore we can directly analyse them.
date_crawled,ad_created & last_seen are represented as string so we need to work on these field before actual analysis



In [9]:
#Analyzing columns represented as string type

autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [10]:
#Analyze date_crawled column
print(autos['date_crawled'].str[:10].unique().shape)
print(autos['date_crawled'].str[:10].value_counts(normalize = True,dropna= False))
print(autos['date_crawled'].str[:10].value_counts(normalize = True,dropna= False).sort_index())

(34,)
2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.

From the above 2 outcomes we can conclude that 
1) The dates between which the crawler
crawled the website is between 2016-03-05  to 2016-04-07
2) Most of the ads were crawled on 2016-04-03 


In [11]:
#Analyze last_seen column

print(autos['last_seen'].str[:10].unique().shape)
print(autos['last_seen'].str[:10].value_counts(normalize = True,dropna = False))
print(autos['last_seen'].str[:10].value_counts(normalize = True,dropna= False).sort_index())

(34,)
2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-12    0.023783
2016-03-31    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023

The last 3 days shows an disporportional amount of change in last_seen data.Its unlikely that the car sales hike disproportionally on these days,Moreover it could be because of end of crawling period.


In [12]:
#Analyse ad_created column
print(autos['ad_created'].str[:10].unique().shape)
print(autos['ad_created'].str[:10].value_counts(normalize = True,dropna = False))
print(autos['ad_created'].str[:10].value_counts(normalize = True,dropna =False).sort_index())

(76,)
2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
2016-03-14    0.035190
2016-04-02    0.035149
2016-03-28    0.034984
2016-03-07    0.034737
2016-03-29    0.034037
2016-03-15    0.034016
2016-03-19    0.033687
2016-04-01    0.033687
2016-03-30    0.033501
2016-03-08    0.033316
2016-03-09    0.033151
2016-03-11    0.032904
2016-03-22    0.032801
2016-03-26    0.032266
2016-03-23    0.032060
2016-03-10    0.031895
2016-03-31    0.031875
2016-03-25    0.031751
2016-03-17    0.031278
2016-03-27    0.030989
2016-03-16    0.030125
2016-03-24    0.029280
2016-03-05    0.022897
2016-03-13    0.017008
2016-03-06    0.015320
                ...   
2016-02-05    0.000041
2016-02-12    0.000041
2016-02-18    0.000041
2016-02-26    0.000041
2016-01-10    0.000041
2016-02-24    0.000041
2016-02-02    0.000041
2016-02-20    0.000041
2016-02-14    0.000041
2016-01-14    0.000021
2016-02-17    0.000021
2016-02-08    0.000021
2015-

We can say that ad were created much before the crawling starts 
The first ad was created on 2015-06-11 & ends on 2016-04-07

In [13]:
#Analyse registration_year column
print(autos['registration_year'].describe())


count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


We noticed some odd values here 
min indicating 1000 and max indicating 9999 which is not an apt year according to our datasets.


In [14]:
#cleaning the Incorrect registration year column data
print(autos['registration_year'].value_counts().sort_index())


1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64


In [15]:
#Ananlysing effect of removal
(~autos["registration_year"].between(1900,2016)).sum()/autos.shape[0]


0.038793369710697

We can say that approximately 4% of values are above & below our listing so we can safely remove them

In [16]:
#Removing unapt rows
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

It appears that most of the vehicles first registered in the past 20 years.

In [17]:
#Exploring by brand names
brands = autos['brand'].value_counts(normalize = True)
top_brands = brands[brands > 0.05].index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

While exploring the brands column we ended up with few brand names which we can choose to further take our experiment with as they all have more than 5% of total brand value.
Let's aggregrate these values to get more insights about our dataset


In [18]:
brand_price_mean ={}
for br in top_brands:
    brand_bool = autos[autos['brand'] == br]
    brand_mean_price = int(brand_bool['price'].mean())
    brand_price_mean[br] = brand_mean_price
brand_price_mean

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

From above results we can conclude that mean prices for brands like **audi**,**bmw** & **mercedes_benz** is way higher than brands like **ford**,**opel**.
brand like **volkswagen** is at intermediate level that has popularity in both worlds of cars.



In [19]:
#Analysing mileage for top 6 brands

brand_mileage_mean = {}
for br in top_brands:
    brand_bool = autos[autos['brand'] == br]
    brand_mean_mil = int(brand_bool['odometer_km'].mean())
    brand_mileage_mean[br] = brand_mean_mil
brand_mileage_mean

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [20]:
#converting both the dictionary (mileage & price) to pandas dataframe to compare them

bmeanprice_series = pd.Series(brand_price_mean).sort_values()
bmeanmileage_series = pd.Series(brand_mileage_mean).sort_values()
df = pd.DataFrame(bmeanprice_series,columns=['mean_price'])
df['mean_mileage'] = bmeanmileage_series
df

mean_price  mean_mileage
opel                 2975        129310
ford                 3749        124266
volkswagen           5402        128707
bmw                  8332        132572
mercedes_benz        8628        130788
audi                 9336        129157

From above conclusion, we can conclude that opel is the best car in terms of price & mileage 
N second conclusion could be that there is not much variation in terms of price & mileage variation.For instance if we compare open & audi which is the top most & last element in our dataframe having high variation in price but it doesn't affect mileage at same proportion.
So we could say that price & mileage are not interlinked closely.

In [21]:
cols = autos.columns
for eachcol in cols:
    if(autos[eachcol].dtype == object):
        print(autos[eachcol].value_counts())
        
        

        

   
    

2016-03-19 17:36:18    3
2016-04-02 11:37:04    3
2016-03-22 09:51:06    3
2016-03-09 11:54:38    3
2016-03-23 18:39:34    3
2016-03-08 10:40:35    3
2016-03-30 19:48:02    3
2016-03-29 23:42:13    3
2016-04-02 15:49:30    3
2016-04-04 16:40:33    3
2016-03-25 19:57:10    3
2016-03-16 21:50:53    3
2016-03-05 16:57:05    3
2016-03-23 19:38:20    3
2016-03-14 20:50:02    3
2016-03-15 19:49:21    2
2016-04-03 04:36:23    2
2016-04-03 11:38:04    2
2016-03-07 15:39:51    2
2016-04-01 20:46:54    2
2016-03-23 22:50:56    2
2016-03-20 18:36:22    2
2016-03-23 14:50:30    2
2016-03-30 12:57:17    2
2016-03-15 19:25:25    2
2016-03-20 22:55:19    2
2016-03-22 15:45:05    2
2016-03-07 21:51:03    2
2016-03-22 16:51:16    2
2016-04-04 12:50:19    2
                      ..
2016-03-10 11:06:21    1
2016-03-25 15:54:39    1
2016-03-08 17:37:39    1
2016-03-25 20:51:02    1
2016-03-25 17:49:12    1
2016-03-14 20:46:40    1
2016-03-17 13:25:20    1
2016-03-29 16:45:11    1
2016-03-12 20:55:28    1


In [22]:
#Replacing german values into it's english counterparts in gearbox column
mapping_dict_gearbox = {
    'manuell' : 'manual',
    'automatik' : 'automatic'
}
autos["gearbox"] = autos["gearbox"].map(mapping_dict_gearbox)
print(autos['gearbox'].value_counts())

manual       34715
automatic     9856
Name: gearbox, dtype: int64


In [23]:
#Replacing german values into it's english counterparts in fuel_type column
mapping_dict_fuel_type = {
    'benzin' : 'benzine',
    'lpg' : 'LPG',
    'cng' : 'CNG',
    'elektro': 'electric',
    'andere': 'other',
    'diesel': 'diesel',
    'hybrid': 'hybrid'
}
autos["fuel_type"] = autos["fuel_type"].map(mapping_dict_fuel_type)
print(autos['fuel_type'].value_counts())

benzine     28540
diesel      14032
LPG           649
CNG            71
hybrid         37
electric       19
other          15
Name: fuel_type, dtype: int64


In [24]:
#Replacing german values into it's english counterparts in unrepaired_damage column
# print(autos['unrepaired_damage'].unique())
mapping_dict_unrepaired_damage = {
    'nein': 'No',
    'ja': 'Yes'
}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(mapping_dict_unrepaired_damage)
print(autos['unrepaired_damage'].value_counts())

No     33834
Yes     4540
Name: unrepaired_damage, dtype: int64


In [ ]:
#converting dates col's datatype from string to int

autos['date_crawled'] = autos['date_crawled'].str.replace("-","")

autos['last_seen'] = autos['last_seen'].str.replace("-","")
autos['ad_created'] = autos['ad_created'].str.replace("-","")
print(autos['ad_created'].head(25))
 



